### ISSUE TITLE
dasdawd123

### ISSUE THREAD

**Issue by davalillo-rm** [davalillo-rm/davalillo-rm#148]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA

print("Iniciando Laboratorio PCA para Curva CDBCRP...")

# --- 1. CARGA DE DATOS ---
# Cambia la ruta a donde guardaste tu base unificada
ruta = r'C:\Ruta\De\Tu\Base_WIDE_Fuente_PCA.xlsx'
df = pd.read_excel(ruta)

# --- 2. AISLAR LA CURVA DE MERCADO ---
# Para el PCA, solo nos interesan las tasas reales de mercado de hoy
cols_tasas = [col for col in df.columns if col.endswith('_TASA_BASE')]
df_tasas = df[['FECHA'] + cols_tasas].set_index('FECHA')

# --- 3. ESTACIONARIEDAD (El secreto del PCA) ---
# En tesorería, el PCA NUNCA se hace sobre las tasas absolutas (porque tienen tendencia).
# Se hace sobre los cambios diarios. Lo multiplicamos por 10,000 para leerlo en Puntos Básicos (bps).
df_diff = df_tasas.diff().dropna() * 10000

# --- 4. ENTRENAMIENTO DEL MODELO PCA ---
# Le pedimos exactamente 3 fuerzas: Nivel, Pendiente y Curvatura
pca = PCA(n_components=3)
pca.fit(df_diff)

# Varianza Explicada (¿Qué tanto poder tiene cada componente?)
varianza = pca.explained_variance_ratio_ * 100
print(f"\nPoder Explicativo del Mercado:")
print(f"PC1 (Nivel): {varianza[0]:.2f}%")
print(f"PC2 (Pendiente): {varianza[1]:.2f}%")
print(f"PC3 (Curvatura): {varianza[2]:.2f}%")
print(f"Total Explicado: {sum(varianza):.2f}%")

# --- 5. VISUALIZACIÓN DE LOS 'LOADINGS' (La prueba de fuego) ---
# Extraemos los vectores (cómo reacciona cada nodo de la curva a cada componente)
loadings = pca.components_

nodos_labels = [col.replace('_TASA_BASE', '') for col in cols_tasas]
x = np.arange(len(nodos_labels))

plt.figure(figsize=(10, 6))
plt.plot(x, loadings[0], marker='o', label=f'PC1 (Nivel) - {varianza[0]:.1f}%', linewidth=2)
plt.plot(x, loadings[1], marker='s', label=f'PC2 (Pendiente) - {varianza[1]:.1f}%', linewidth=2)
plt.plot(x, loadings[2], marker='^', label=f'PC3 (Curvatura) - {varianza[2]:.1f}%', linewidth=2)

plt.axhline(0, color='black', linestyle='--', linewidth=1)
plt.xticks(x, nodos_labels)
plt.title('Vectores Propios del PCA (Sensibilidad de la Curva CDBCRP)', fontsize=14)
plt.xlabel('Nodos de Maduración')
plt.ylabel('Impacto Direccional')
plt.legend()
plt.grid(alpha=0.3)
plt.tight_layout()
plt.show()
